components/scream/src/diagnostics/tests/atm_density_test.cpp

```diff
@@ -0,0 +1,187 @@
+          pressure("pressure",num_mid_packs),
+          watervapor("watervapor",num_mid_packs);
+
+  auto dview_as_real = [&] (const view_1d& v) -> rview_1d {
+    return rview_1d(reinterpret_cast<Real*>(v.data()),v.size()*packsize);
+  };
+
+  // Construct random input data
+  using RPDF = std::uniform_real_distribution<Real>;
+  RPDF pdf_qv(1e-6,1e-3),
+       pdf_pseudodens(1.0,100.0),
+       pdf_pres(0.0,PC::P0),
+       pdf_temp(200.0,400.0);
+
+  // A time stamp
+  util::TimeStamp t0 ({2022,1,1},{0,0,0});
+
+  // Construct the Diagnostic
+  ekat::ParameterList params;
+  params.set<std::string>("Diagnostic Name", "Atmosphere Density");
+  params.set<std::string>("Grid", "Point Grid");
+  register_diagnostics();
+  auto& diag_factory = AtmosphereDiagnosticFactory::instance();
+  auto diag = diag_factory.create("AtmosphereDensity",comm,params);
+  diag->set_grids(gm);
+
+
+  // Set the required fields for the diagnostic.
+  std::map<std::string,Field> input_fields;
+  for (const auto& req : diag->get_required_field_requests()) {
+    Field f(req.fid);
+    auto & f_ap = f.get_header().get_alloc_properties();
+    f_ap.request_allocation(packsize);
+    f.allocate_view();
+    const auto name = f.name();
+    f.get_header().get_tracking().update_time_stamp(t0);
+    diag->set_required_field(f.get_const());
+    REQUIRE_THROWS(diag->set_computed_field(f));
+    input_fields.emplace(name,f);
+  }
+
+  // Initialize the diagnostic
+  diag->initialize(t0,RunType::Initial);
+
+  // Run tests
+  {
+    // Construct random data to use for test
+    // Get views of input data and set to random values
+    const auto& T_mid_f       = input_fields["T_mid"];
+    const auto& T_mid_v       = T_mid_f.get_view<Pack**>();
+    const auto& pseudo_dens_f = input_fields["pseudo_density"];
+    const auto& pseudo_dens_v = pseudo_dens_f.get_view<Pack**>();
+    const auto& p_mid_f       = input_fields["p_mid"];
+    const auto& p_mid_v       = p_mid_f.get_view<Pack**>();
+    const auto& qv_mid_f      = input_fields["qv"];
+    const auto& qv_mid_v      = qv_mid_f.get_view<Pack**>();
+    for (int icol = 0; icol<ncols;++icol) {
+      const auto& T_sub      = ekat::subview(T_mid_v,icol);
+      const auto& pseudo_sub = ekat::subview(pseudo_dens_v,icol);
+      const auto& p_sub      = ekat::subview(p_mid_v,icol);
+      const auto& qv_sub     = ekat::subview(qv_mid_v,icol);
+      ekat::genRandArray(dview_as_real(temperature),   engine, pdf_temp);
+      ekat::genRandArray(dview_as_real(pseudodensity), engine, pdf_pseudodens);
+      ekat::genRandArray(dview_as_real(pressure),      engine, pdf_pres);
+      ekat::genRandArray(dview_as_real(watervapor),    engine, pdf_qv);
+      Kokkos::deep_copy(T_sub,temperature);
+      Kokkos::deep_copy(pseudo_sub,pseudodensity);
+      Kokkos::deep_copy(p_sub,pressure);
+      Kokkos::deep_copy(qv_sub,watervapor);
+    } 
+
+    // Run diagnostic and compare with manual calculation
```

Comment:
I am not sure I like this kind of test. It's basically a lot of boiler plate code, and the important part is kind of a copy paste of the same code in the diagnostic run method. It's somewhat of a burden in terms of lines of code to maintain vs lines of code tested.

Alas, I'm not really sure how to test all these diagnostics classes, which do come down to a lot of boilerplate code that calls a 5 lines function...

Reply:
I agree, my hope is that we won't have to update all the tests at once in the future.  So the repeated boilerplate code won't be too much of a burden.  I also struggled with how to best test these new diagnostics.  I think they are a special case though, since they are just output from defined physics functions, i.e. they are not new functions themselves.  Since the common physics functions are, themselves, robustly tested anyway I didn't see a need to replicate those tests here.  Instead, if the diagnostic produces the same output as the common physics function does then the diagnostic is doing exactly what we want it too.

Other diagnostics should be more rigorously unit tested.